src/ripple/app/hook/impl/applyHook.cpp

```diff
@@ -4988,33 +7500,94 @@ DEFINE_HOOK_FUNCTION(
         man1 *= -1LL;
 
     return make_float((uint64_t)man1, exp1, is_negative(float1));
+}
+
+DEFINE_WASM_FUNCTION(
+    int64_t,
+    float_mulratio,
+    int64_t float1,
+    uint32_t round_up,
+    uint32_t numerator,
+    uint32_t denominator)
+{
+    WASM_HOOK_SETUP();  // populates memory_ctx, memory, memory_length,
+                        // applyCtx, hookCtx on current stack
+
+    return __float_mulratio(
+        hookCtx, applyCtx, j, float1, round_up, numerator, denominator);
+
+    WASM_HOOK_TEARDOWN();
+}
+
+DEFINE_JS_FUNCTION(
+    JSValue,
+    float_mulratio,
+    JSValue raw_f1,
+    JSValue raw_ru,
+    JSValue raw_n,
+    JSValue raw_d)
+{
+    JS_HOOK_SETUP();
+
+    auto [f1, ru, n, d] = FromJSInts(ctx, raw_f1, raw_ru, raw_n, raw_d);
 
-    HOOK_TEARDOWN();
+    if (any_missing(f1, ru, n, d) || !fits_u32(ru, n, d))
+        returnJS(INVALID_ARGUMENT);
+
+    int64_t const out = __float_mulratio(
+        hookCtx,
+        applyCtx,
+        j,
+        *f1,
+        (uint32_t)(*ru),
+        (uint32_t)(*n),
+        (uint32_t)(*d));
+    if (out < 0)
+        returnJS(out);
+
+    returnJSXFL(out);
+
+    JS_HOOK_TEARDOWN();
 }
 
-DEFINE_HOOK_FUNCTION(int64_t, float_negate, int64_t float1)
+DEFINE_WASM_FUNCTION(int64_t, float_negate, int64_t float1)
 {
-    HOOK_SETUP();  // populates memory_ctx, memory, memory_length, applyCtx,
-                   // hookCtx on current stack
+    WASM_HOOK_SETUP();  // populates memory_ctx, memory, memory_length,
+                        // applyCtx, hookCtx on current stack
 
     if (float1 == 0)
         return 0;
     RETURN_IF_INVALID_FLOAT(float1);
     return hook_float::invert_sign(float1);
 
-    HOOK_TEARDOWN();
+    WASM_HOOK_TEARDOWN();
 }
 
-DEFINE_HOOK_FUNCTION(
-    int64_t,
-    float_compare,
+DEFINE_JS_FUNCTION(JSValue, float_negate, JSValue raw_f1)
+{
+    JS_HOOK_SETUP();
+
+    auto float1 = FromJSInt(ctx, raw_f1);
+    if (!float1.has_value())
+        returnJS(INVALID_ARGUMENT);
+
+    if (*float1 == 0)
+        returnJS(0);
```

Comment:
The success value should be returned by XFL.

```suggestion
        returnJSXFL(0);
```

Reply:
https://github.com/Xahau/xahaud/pull/481